rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,49 @@
+use crate::memory::{ic::NEXT_REGION_ID, Memory};
+use crate::types::{size_of, Region, Value, Words, TAG_REGION};
+use crate::rts_trap_with;
+
+use motoko_rts_macros::ic_mem_fn;
+
+#[ic_mem_fn]
+pub unsafe fn region_new<M: Memory>(mem: &mut M) -> Value {
+    let r_ptr = mem.alloc_words(size_of::<Region>() + Words(1));
+    // NB. cannot use as_region() here as we didn't write the header yet
+    let region = r_ptr.get_ptr() as *mut Region;
+    (*region).header.tag = TAG_REGION;
+    (*region).id = NEXT_REGION_ID;
+    NEXT_REGION_ID += 1;
+    (*region).page_count = 0;
+    Value::from_ptr(region as usize)
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_id<M: Memory>(_mem: &mut M, r: Value) -> Value {
+    let r = r.as_region();
+    Value::from_raw((*r).id as u32) // is this right?  It makes the test pass.
```

Comment:
@crusso @luc-blaeser This is the use of `from_raw` that *feels* potentially wrong, since I don't understand the conventions yet.

Reply:
Basically object pointers are represented "skewed" by pointer -1 so always end in 11 (lowest two bits set) cos word aligned to start with. nat8 and nat16 are shifted left 16 bits. Other unboxed scalar values are in top 30 bits. True is just 1. Not sure about false, probably 0. Options are special.

Reply:
What is the return value supposed to be, a Nat16? Then I expect you need to shift left by 16 bits and return that. See TaggedSmallWord.msb_adjust in compile.ml

Reply:
Ah, sorry, I now see that Prim.regionId returns a Nat32, not Nat16. In that case, I think its just suffiecient to call Value.from_scalar(*r).id as u32) which shifts left by 1. I believe small  Nat32's are unboxed.

Reply:
Ok, I think you need to decide if these operations are returning (Vanilla) Motoko values, or just u32s, and have the compiler Prim implementations, that call out to these, behave accordingly.